onnxruntime/contrib_ops/webgpu/bert/group_query_attention.cc

```diff
@@ -0,0 +1,99 @@
+
+  GroupQueryAttentionParameters params;
+  ORT_RETURN_IF_ERROR(group_query_attention_helper::CheckInputs(query,
+                                                                key,
+                                                                value,
+                                                                past_key,
+                                                                past_value,
+                                                                cos_cache,
+                                                                sin_cache,
+                                                                &params,
+                                                                num_heads_,
+                                                                kv_num_heads_,
+                                                                seqlen_k,
+                                                                total_seqlen_tensor,
+                                                                scale_,
+                                                                softcap_));
+  WebgpuAttentionParameters parameters(params);
+  if (parameters.is_packed_qkv_) {
+    ORT_NOT_IMPLEMENTED("Packed QKV of shape (B, L, N, 3, H) not implemented for webgpu-ep.");
+  }
+  TensorShapeVector output_shape(3);
+  output_shape[0] = static_cast<int64_t>(parameters.batch_size_);
+  output_shape[1] = static_cast<int64_t>(parameters.sequence_length_);
+  output_shape[2] = static_cast<int64_t>(parameters.v_hidden_size_);
+  Tensor* output = context.Output(0, output_shape);
+  const int present_kv_seqlen = parameters.seqlen_present_kv_cache_;
+  std::vector<int64_t> present_kv_shape({static_cast<int64_t>(parameters.batch_size_), static_cast<int64_t>(kv_num_heads_), static_cast<int64_t>(present_kv_seqlen), static_cast<int64_t>(parameters.head_size_)});
+  Tensor* present_key = context.Output(1, present_kv_shape);
+  Tensor* present_value = context.Output(2, present_kv_shape);
+
+  TensorShapeVector q_new_dims({parameters.batch_size_, parameters.num_heads_,
+                                parameters.sequence_length_, parameters.head_size_});
+  TensorShape q_new_shape(q_new_dims);
+  Tensor Q = context.CreateGPUTensor(query->DataType(), q_new_shape);
+  ORT_RETURN_IF_ERROR(TransferBSDToBNSH(
+      context, parameters.num_heads_, parameters.sequence_length_, parameters.head_size_, query, nullptr, 0, &Q));
+  TensorShapeVector k_new_dims({parameters.batch_size_, parameters.num_heads_,
+                                parameters.kv_sequence_length_, parameters.head_size_});
+  if (parameters.qkv_format_ == Q_K_V_BSNH_BNSH_BNSH) {  // key and value in BNSH format
+    return ApplyAttention(&Q, key, value, nullptr, past_key, past_value, output, present_key,
+                          present_value, parameters, context, seqlen_k, total_seqlen_tensor);
+  }
+  TensorShape k_new_shape(k_new_dims);
+  Tensor K = context.CreateGPUTensor(key->DataType(), k_new_shape);
```

Comment:
This line causes a segfault with these models: https://huggingface.co/microsoft/Phi-3-mini-4k-instruct-onnx/tree/main/cpu_and_mobile as they have GQA nodes that do not have the optional `key` and `value` inputs so the Tensor* is a nullptr.

https://github.com/microsoft/onnxruntime/blob/c64459fdd7bed84340d98d58daffb736151b7b78/onnxruntime/core/graph/contrib_ops/bert_defs.cc#L1090-L1111